### PR TITLE
rsa bindings

### PR DESCRIPTION
**Pull request by DmitriyMusatkin**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/crypto.c [hunk 1]
```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
...
+    int encrypt_algo = 0;
+    const char *plaintext_ptr;
+    Py_ssize_t plaintext_len;
+    if (!PyArg_ParseTuple(args, "Ois#", &rsa_capsule, &encrypt_algo, &plaintext_ptr, &plaintext_len)) {
+        return NULL;
+    }
+
+    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
+    if (rsa == NULL) {
+        return PyErr_AwsLastError();
```
Comment (graebm) [inline source/crypto.c:369]: PyCapsule_GetPointer() already sets a good python exception if it fails. Calling PyErr_AwsLastError() will overwrite it with nonsense from whatever stale value aws_last_error() returns.
```suggestion
    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
    if (rsa == NULL) {
        return NULL;
```

File: source/crypto.c [hunk 2]
```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
...
+    int encrypt_algo = 0;
+    const char *ciphertext_ptr;
+    Py_ssize_t ciphertext_len;
+    if (!PyArg_ParseTuple(args, "Ois#", &rsa_capsule, &encrypt_algo, &ciphertext_ptr, &ciphertext_len)) {
+        return NULL;
+    }
+
+    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
+    if (rsa == NULL) {
+        return PyErr_AwsLastError();
```
Comment (graebm) [inline source/crypto.c:399]: `return NULL`

File: source/crypto.c [hunk 3]
```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
...
+    }
+
+    struct aws_byte_cursor ciphertext_cur = aws_byte_cursor_from_array(ciphertext_ptr, ciphertext_len);
+    struct aws_byte_buf result_buf;
+    aws_byte_buf_init(&result_buf, allocator, aws_rsa_key_pair_block_length(rsa));
+
+    if (aws_rsa_key_pair_decrypt(rsa, encrypt_algo, ciphertext_cur, &result_buf)) {
+        return PyErr_AwsLastError();
```
Comment (graebm) [inline source/crypto.c:407]: leaking `result_buf`

File: source/crypto.c [hunk 4]
```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
...
+    int sign_algo = 0;
+    const char *digest_ptr;
+    Py_ssize_t digest_len;
+    if (!PyArg_ParseTuple(args, "Ois#", &rsa_capsule, &sign_algo, &digest_ptr, &digest_len)) {
+        return NULL;
+    }
+
+    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
+    if (rsa == NULL) {
+        return PyErr_AwsLastError();
```
Comment (graebm) [inline source/crypto.c:429]:
```suggestion
        return NULL;
```

File: source/crypto.c [hunk 5]
```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
...
+    }
+
+    struct aws_byte_cursor digest_cur = aws_byte_cursor_from_array(digest_ptr, digest_len);
+    struct aws_byte_buf result_buf;
+    aws_byte_buf_init(&result_buf, allocator, aws_rsa_key_pair_signature_length(rsa));
+
+    if (aws_rsa_key_pair_sign_message(rsa, sign_algo, digest_cur, &result_buf)) {
+        return PyErr_AwsLastError();
```
Comment (graebm) [inline source/crypto.c:437]: leaking `result_buf`

File: source/module.c [hunk 6]
```diff
@@ -18,6 +18,7 @@
 #include "websocket.h"
 
 #include <aws/auth/auth.h>
+#include <aws/cal/cal.h>
```
Comment (graebm) [inline source/module.c:21]: utterly trivial: why was this added?
```suggestion
```

File: test/test_crypto.py [hunk 7]
```diff
@@ -76,6 +116,62 @@ def test_md5_iterated(self):
...
+        digest = h.digest()
+
+        param_list = [RSASignatureAlgorithmType.PKCS1_5_SHA256,
+                      RSASignatureAlgorithmType.PSS_SHA256]
+
+        for p in param_list:
+            with self.subTest(msg="RSA Signing Roundtrip using algo p", p=p):
+                rsa = RSA.rsa_private_key_from_pem_data(RSA_PRIVATE_KEY_PEM)
+                signature = rsa.sign(p, digest)
+                self.assertEqual(rsa.verify(p, digest, signature), True)
```
Comment (graebm) [inline test/test_crypto.py:149]: utterly trivial: 
```suggestion
                self.assertTrue(rsa.verify(p, digest, signature))
```

File: test/test_crypto.py [hunk 8]
```diff
@@ -76,6 +116,62 @@ def test_md5_iterated(self):
...
+
+                rsa_pub = RSA.rsa_private_key_from_pem_data(RSA_PRIVATE_KEY_PEM)
+                self.assertEqual(rsa_pub.verify(p, digest, signature), True)
```
Comment (graebm) [inline test/test_crypto.py:152]:
```suggestion
                self.assertTrue(rsa_pub.verify(p, digest, signature))
```

File: test/test_crypto.py [hunk 9]
```diff
@@ -76,6 +116,62 @@ def test_md5_iterated(self):
...
+        digest = h.digest()
+
+        h2 = Hash.sha256_new()
+        h2.update(b'another totally original test string')
+        digest2 = h2.digest()
+
+        rsa = RSA.rsa_private_key_from_pem_data(RSA_PRIVATE_KEY_PEM)
+        signature = rsa.sign(RSASignatureAlgorithmType.PKCS1_5_SHA256, digest)
+        self.assertEqual(rsa.verify(RSASignatureAlgorithmType.PKCS1_5_SHA256, digest2, signature), False)
+        self.assertEqual(rsa.verify(RSASignatureAlgorithmType.PKCS1_5_SHA256, digest, b'bad signature'), False)
```
Comment (graebm) [inline test/test_crypto.py:173]:
```suggestion
        self.assertFalse(rsa.verify(RSASignatureAlgorithmType.PKCS1_5_SHA256, digest2, signature))
        self.assertFalse(rsa.verify(RSASignatureAlgorithmType.PKCS1_5_SHA256, digest, b'bad signature'))
```

File: awscrt/crypto.py [hunk 10]
```diff
@@ -59,3 +61,91 @@ def update(self, to_hmac):
...
+    @staticmethod
+    def rsa_public_key_from_pem_data(pem_data: Union[str, bytes, bytearray, memoryview]):
+        """
+        Creates a new instance of public RSA key pair from pem data.
+        Raises ValueError if pem does not have public key object.
+        """
+        return RSA(native_handle=_awscrt.rsa_public_key_from_pem_data(pem_data))
+
+    def encrypt(self, encryption_algorithm: RSAEncryptionAlgorithmType,
+                plaintext: Union[str, bytes, bytearray, memoryview]) -> bytes:
```
[the remark this answers is not in the document]
Comment (DmitriyMusatkin) [reply awscrt/crypto.py:123]: str probably makes sense for plaintext. but for ciphertext i agree removing str makes more sense. updating it

File: source/crypto.c (hunk 1 above)
Comment (DmitriyMusatkin) [reply source/crypto.c:369]: ah, was copying the behavior from sha impl earlier in the file. will fix

File: source/module.c (hunk 6 above)
Comment (DmitriyMusatkin) [reply source/module.c:21]: ah, can be removed. was experimenting with error mapping and forgot to get rid of it

File: awscrt/crypto.py [hunk 11]
```diff
@@ -59,3 +61,91 @@ def update(self, to_hmac):
...
+        return _awscrt.rsa_encrypt(self._rsa, encryption_algorithm, plaintext)
+
+    def decrypt(self, encryption_algorithm: RSAEncryptionAlgorithm,
+                ciphertext: Union[bytes, bytearray, memoryview]) -> bytes:
+        """
+        Decrypts data using a given algorithm.
+        """
+        return _awscrt.rsa_decrypt(self._rsa, encryption_algorithm, ciphertext)
+
+    def sign(self, encryption_algorithm: RSASignatureAlgorithm,
```
Comment (graebm) [inline awscrt/crypto.py:136]: is this right???
```suggestion
    def sign(self, signature_algorithm: RSASignatureAlgorithm,
```

File: awscrt/crypto.py [hunk 12]
```diff
@@ -59,3 +61,91 @@ def update(self, to_hmac):
...
+             digest: Union[bytes, bytearray, memoryview]) -> bytes:
+        """
+        Signs data using a given algorithm.
+        Note: function expects digest of the message, ex sha256
+        """
+        return _awscrt.rsa_sign(self._rsa, encryption_algorithm, digest)
+
+    def verify(self, encryption_algorithm: RSASignatureAlgorithm,
```
Comment (graebm) [inline awscrt/crypto.py:144]: same name/type mismatch

**Review by graebm — APPROVED** [review]
fix & ship
